lib/active_model/serializer/adapter/json_sideload.rb

```diff
@@ -0,0 +1,185 @@
+module ActiveModel
+  class Serializer
+    class Adapter
+      class JsonSideload < Json
+
+        # the list of what has already been serialized will be kept here to
+        # help avoid infinite recursion
+        #
+        # this should be a list of association_name to a list of objects
+        attr_accessor :associations
+
+        # When we are sideloading associations, we can more easily track what
+        # has been serialized, so that we avoid infinite
+        # recursion / serialization.
+        def initialize(*args)
+          @associations = {}
+
+          super(*args)
+        end
+
+        def serializable_hash(options = nil)
+          options ||= {}
+
+          serialize_hash(options)
+
+          singularize_lone_associations
+
+          @associations
+        end
```

Comment:
I'm ok with this as a way to get it out the door in a kind of feature toggle, but it should probably be generalized so that either json or jsonapi could use it, no?